File: experiment1.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

# STEP 1: Read Images
# Foreground image (Tom) with alpha channel
fg_rgba = cv2.imread('input/tom.png', cv2.IMREAD_UNCHANGED)
if fg_rgba is None:
    raise FileNotFoundError("input/tom.png not found")

# Extract BGR channels and Alpha channel
if fg_rgba.shape[2] == 4:
    fg = fg_rgba[:, :, :3]
    alpha = fg_rgba[:, :, 3]
else:
    fg = fg_rgba
    alpha = np.ones(fg.shape[:2], dtype=np.uint8) * 255

# Create and save the binary mask
_, mask = cv2.threshold(alpha, 128, 255, cv2.THRESH_BINARY)
cv2.imwrite('output/mask.png', mask)

# Background image
bg = cv2.imread('input/background.jpg')
if bg is None:
    raise FileNotFoundError("input/background.jpg not found")

# STEP 2: Resize Images to Match Foreground Dimensions
rows, cols = fg.shape[:2]
bg = cv2.resize(bg, (cols, rows))
mask = cv2.resize(mask, (cols, rows))

# STEP 3: Convert Mask to Binary (0 or 1 range)
mask_normalized = mask.astype(np.float64) / 255.0

# STEP 4: Convert Images to Double Precision (0 to 1 range)
fg_double = fg.astype(np.float64) / 255.0
bg_double = bg.astype(np.float64) / 255.0

# STEP 5: Create 3-Channel Mask
mask3 = np.stack([mask_normalized]*3, axis=2)

# STEP 6: Extract Foreground
extracted_fg = fg_double * mask3

# STEP 7: Remove Foreground Region from Background
extracted_bg = bg_double * (1.0 - mask3)

# STEP 8: Create Final Composite Image
final_output = extracted_fg + extracted_bg

# Save intermediate outputs
cv2.imwrite('output/extracted_fg.png', (extracted_fg * 255).astype(np.uint8))
cv2.imwrite('output/masked_bg.png', (extracted_bg * 255).astype(np.uint8))
cv2.imwrite('output/composite.png', (final_output * 255).astype(np.uint8))

# STEP 9: Display All Results (Save as matplotlib figure)
# Convert BGR to RGB for matplotlib display
fg_rgb = cv2.cvtColor(fg, cv2.COLOR_BGR2RGB)
bg_rgb = cv2.cvtColor(bg, cv2.COLOR_BGR2RGB)
extracted_fg_rgb = cv2.cvtColor((extracted_fg * 255).astype(np.uint8), cv2.COLOR_BGR2RGB)
extracted_bg_rgb = cv2.cvtColor((extracted_bg * 255).astype(np.uint8), cv2.COLOR_BGR2RGB)
final_output_rgb = cv2.cvtColor((final_output * 255).astype(np.uint8), cv2.COLOR_BGR2RGB)

plt.figure(figsize=(15, 10))
plt.suptitle('Foreground Extraction and Image Compositing', fontsize=16)

plt.subplot(2, 3, 1)
plt.imshow(fg_rgb)
plt.title('Original Foreground (Tom/Jerry)')
plt.axis('off')

plt.subplot(2, 3, 2)
plt.imshow(bg_rgb)
plt.title('Background Image')
plt.axis('off')

plt.subplot(2, 3, 3)
plt.imshow(mask, cmap='gray')
plt.title('Binary Mask')
plt.axis('off')

plt.subplot(2, 3, 4)
plt.imshow(extracted_fg_rgb)
plt.title('Extracted Foreground')
plt.axis('off')

plt.subplot(2, 3, 5)
plt.imshow(extracted_bg_rgb)
plt.title('Masked Background')
plt.axis('off')

plt.subplot(2, 3, 6)
plt.imshow(final_output_rgb)
plt.title('Final Composite Image')
plt.axis('off')

plt.tight_layout()
plt.savefig('output/Experiment1_Results.png', dpi=150)
print("Output successfully generated and saved to output/Experiment1_Results.png!")

File: experiment1.m
% =========================================================
% Experiment 1: Foreground Extraction and Image Compositing
% Using Tom and Jerry Images
% Student: Sanyog Kumar Singh | USN: 23BTRDC034
% Subject: Digital Image Processing (23DSDE12)
% =========================================================

clc;
clear;
close all;

%% STEP 1: Read Images

% Foreground image (Tom or Jerry)
% Check if image is in 'input/' folder (local) or current folder (MATLAB Online)
if exist('input/tom.png', 'file')
    [fg, map, alpha] = imread('input/tom.png');
elseif exist('tom.png', 'file')
    [fg, map, alpha] = imread('tom.png');
else
    error('Unable to find tom.png. Please ensure it is uploaded to MATLAB Online.');
end

% Background image
if exist('input/background.jpg', 'file')
    bg = imread('input/background.jpg');
elseif exist('background.jpg', 'file')
    bg = imread('background.jpg');
else
    error('Unable to find background.jpg. Please ensure it is uploaded to MATLAB Online.');
end

%% STEP 2: Resize Images to Match Foreground Dimensions

[rows, cols, ~] = size(fg);

% Resize background to match foreground
bg = imresize(bg, [rows cols]);

%% STEP 3: Create Binary Mask from Alpha Channel

% If the image has a transparent background (alpha channel), use it as the mask.
if ~isempty(alpha)
    mask = alpha > 128; % Convert alpha to logical binary mask
else
    % Fallback if no transparency: generate auto-mask using intensity
    gray = rgb2gray(fg);
    mask = imbinarize(gray);
    mask = imfill(mask, 'holes');
    mask = bwareaopen(mask, 500);
end

%% STEP 4: Convert Images to Double Precision

fg   = im2double(fg);
bg   = im2double(bg);
mask = im2double(mask);

%% STEP 5: Create 3-Channel (RGB) Mask

% Replicate single-channel mask across all 3 colour channels
mask3 = cat(3, mask, mask, mask);

%% STEP 6: Extract Foreground

% Keep only the foreground object (Tom/Jerry) using the mask
extracted_fg = fg .* mask3;

%% STEP 7: Remove Foreground Region from Background

% Cut out the region where Tom/Jerry will be placed
extracted_bg = bg .* (1 - mask3);

%% STEP 8: Create Final Composite Image

% Combine extracted foreground + masked background
final_output = extracted_fg + extracted_bg;

%% STEP 9: Display All Results

figure('Name', 'Foreground Extraction and Image Compositing');

subplot(2, 3, 1);
imshow(fg);
title('Original Foreground (Tom/Jerry)');

subplot(2, 3, 2);
imshow(bg);
title('Background Image');

subplot(2, 3, 3);
imshow(mask);
title('Binary Mask');

subplot(2, 3, 4);
imshow(extracted_fg);
title('Extracted Foreground');

subplot(2, 3, 5);
imshow(extracted_bg);
title('Masked Background');

subplot(2, 3, 6);
imshow(final_output);
title('Final Composite Image');

%% STEP 10: Save Outputs to Folder

% Create 'output' directory if it doesn't exist
if ~exist('output', 'dir')
    mkdir('output');
end

% Save outputs to the 'output' directory
imwrite(mask, 'output/mask.png');
imwrite(extracted_fg, 'output/extracted_fg.png');
imwrite(extracted_bg, 'output/masked_bg.png');
imwrite(final_output, 'output/composite.png');

% Save the entire figure
set(gcf, 'PaperPositionMode', 'auto');
print('output/Experiment1_Results','-dpng','-r0');
